File: FileServerAndroid/app/src/main/java/com/syj5385/fileexplore/MainActivity.java
package com.syj5385.fileexplore;

import android.app.Activity;
import android.app.AlertDialog;
import android.app.DownloadManager;
import android.content.ActivityNotFoundException;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Bitmap;
import android.net.Uri;
import android.net.http.SslError;
import android.os.Bundle;
import android.os.Environment;
import android.view.LayoutInflater;
import android.view.View;
import android.webkit.CookieManager;
import android.webkit.DownloadListener;
import android.webkit.SslErrorHandler;
import android.webkit.URLUtil;
import android.webkit.ValueCallback;
import android.webkit.WebChromeClient;
import android.webkit.WebResourceError;
import android.webkit.WebResourceRequest;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import java.net.URISyntaxException;
import java.util.Locale;

public final class MainActivity extends Activity {
    private static final String PREFS_NAME = "file_explore_preferences";
    private static final String PREF_SERVER_URL = "server_url";
    private static final int FILE_CHOOSER_REQUEST = 1001;

    private WebView webView;
    private ProgressBar progressBar;
    private LinearLayout errorPanel;
    private TextView errorTitle;
    private TextView errorMessage;
    private TextView serverStatus;
    private SharedPreferences preferences;
    private ValueCallback<Uri[]> pendingFileCallback;
    private Uri configuredServer;
    private String temporarilyTrustedSslHost;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        preferences = getSharedPreferences(PREFS_NAME, MODE_PRIVATE);
        webView = findViewById(R.id.webView);
        progressBar = findViewById(R.id.progressBar);
        errorPanel = findViewById(R.id.errorPanel);
        errorTitle = findViewById(R.id.errorTitle);
        errorMessage = findViewById(R.id.errorMessage);
        serverStatus = findViewById(R.id.serverStatus);

        configureWebView();

        findViewById(R.id.reloadButton).setOnClickListener(view -> reloadServer());
        findViewById(R.id.settingsButton).setOnClickListener(view -> showServerSettings(false));
        findViewById(R.id.retryButton).setOnClickListener(view -> reloadServer());

        String savedUrl = preferences.getString(PREF_SERVER_URL, "");
        configuredServer = parseServerUri(savedUrl);
        updateServerStatus();

        if (savedInstanceState != null && configuredServer != null) {
            webView.restoreState(savedInstanceState);
        } else if (configuredServer != null) {
            webView.loadUrl(configuredServer.toString());
        } else {
            showServerSettings(true);
        }
    }

    private void configureWebView() {
        WebSettings settings = webView.getSettings();
        settings.setJavaScriptEnabled(true);
        settings.setDomStorageEnabled(true);
        settings.setDatabaseEnabled(true);
        settings.setAllowContentAccess(true);
        settings.setAllowFileAccess(false);
        settings.setAllowFileAccessFromFileURLs(false);
        settings.setAllowUniversalAccessFromFileURLs(false);
        settings.setMixedContentMode(WebSettings.MIXED_CONTENT_NEVER_ALLOW);
        settings.setMediaPlaybackRequiresUserGesture(true);
        settings.setSupportMultipleWindows(false);
        settings.setBuiltInZoomControls(false);
        settings.setDisplayZoomControls(false);
        settings.setUseWideViewPort(true);
        settings.setLoadWithOverviewMode(false);
        settings.setSafeBrowsingEnabled(true);
        settings.setUserAgentString(settings.getUserAgentString() + " FileExploreAndroid/1.0");

        CookieManager cookieManager = CookieManager.getInstance();
        cookieManager.setAcceptCookie(true);
        cookieManager.setAcceptThirdPartyCookies(webView, false);

        webView.setWebViewClient(new FileExploreWebViewClient());
        webView.setWebChromeClient(new FileExploreChromeClient());
        webView.setDownloadListener(createDownloadListener());
    }

    private DownloadListener createDownloadListener() {
        return (url, userAgent, contentDisposition, mimeType, contentLength) -> {
            try {
                String fileName = URLUtil.guessFileName(url, contentDisposition, mimeType);
                DownloadManager.Request request = new DownloadManager.Request(Uri.parse(url));
                request.setTitle(fileName);
                request.setDescription("FileExplore에서 다운로드 중");
                request.setMimeType(mimeType);
                request.setNotificationVisibility(
                        DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED
                );
                request.setAllowedOverMetered(true);
                request.setAllowedOverRoaming(false);
                request.setDestinationInExternalPublicDir(Environment.DIRECTORY_DOWNLOADS, fileName);

                String cookies = CookieManager.getInstance().getCookie(url);
                if (cookies != null && !cookies.trim().isEmpty()) {
                    request.addRequestHeader("Cookie", cookies);
                }
                if (userAgent != null && !userAgent.trim().isEmpty()) {
                    request.addRequestHeader("User-Agent", userAgent);
                }

                DownloadManager manager = (DownloadManager) getSystemService(DOWNLOAD_SERVICE);
                manager.enqueue(request);
                Toast.makeText(this, "다운로드를 시작했습니다: " + fileName, Toast.LENGTH_LONG).show();

                Uri downloadUri = Uri.parse(url);
                if (temporarilyTrustedSslHost != null
                        && temporarilyTrustedSslHost.equalsIgnoreCase(downloadUri.getHost())) {
                    Toast.makeText(
                            this,
                            "참고: Android 다운로드 관리자는 자체 서명 인증서를 거부할 수 있습니다.",
                            Toast.LENGTH_LONG
                    ).show();
                }
            } catch (Exception exception) {
                Toast.makeText(this, "다운로드를 시작하지 못했습니다.", Toast.LENGTH_LONG).show();
            }
        };
    }

    private void reloadServer() {
        hideError();
        if (configuredServer == null) {
            showServerSettings(true);
            return;
        }
        if (webView.getUrl() == null) {
            webView.loadUrl(configuredServer.toString());
        } else {
            webView.reload();
        }
    }

    private void showServerSettings(boolean required) {
        View content = LayoutInflater.from(this).inflate(R.layout.dialog_server_settings, null, false);
        EditText input = content.findViewById(R.id.serverUrlInput);
        String current = configuredServer == null ? "" : configuredServer.toString();
        input.setText(current);

        AlertDialog dialog = new AlertDialog.Builder(this)
                .setTitle("FileServer 주소")
                .setView(content)
                .setPositiveButton("저장", null)
                .setNegativeButton(required ? null : "취소", null)
                .create();
        dialog.setCancelable(!required);
        dialog.setCanceledOnTouchOutside(!required);
        dialog.setOnShowListener(ignored -> dialog.getButton(AlertDialog.BUTTON_POSITIVE)
                .setOnClickListener(button -> {
                    Uri parsed = parseServerUri(input.getText().toString());
                    if (parsed == null) {
                        input.setError("http:// 또는 https:// 주소를 입력하세요.");
                        return;
                    }
                    configuredServer = parsed;
                    temporarilyTrustedSslHost = null;
                    preferences.edit().putString(PREF_SERVER_URL, parsed.toString()).apply();
                    updateServerStatus();
                    hideError();
                    webView.clearHistory();
                    webView.loadUrl(parsed.toString());
                    dialog.dismiss();
                }));
        dialog.show();
    }

    private Uri parseServerUri(String value) {
        if (value == null) {
            return null;
        }
        String normalized = value.trim();
        if (normalized.isEmpty()) {
            return null;
        }
        if (!normalized.contains("://")) {
            normalized = "https://" + normalized;
        }
        Uri uri = Uri.parse(normalized);
        String scheme = uri.getScheme();
        if (uri.getHost() == null
                || scheme == null
                || !(scheme.equalsIgnoreCase("http") || scheme.equalsIgnoreCase("https"))) {
            return null;
        }
        String text = uri.toString();
        while (text.endsWith("/") && text.length() > scheme.length() + 3) {
            text = text.substring(0, text.length() - 1);
        }
        return Uri.parse(text);
    }

    private void updateServerStatus() {
        if (configuredServer == null) {
            serverStatus.setText(R.string.server_not_configured);
            return;
        }
        String host = configuredServer.getHost();
        int port = configuredServer.getPort();
        serverStatus.setText(port > 0 ? host + ":" + port : host);
    }

    private boolean belongsToConfiguredServer(Uri uri) {
        return configuredServer != null
                && uri != null
                && configuredServer.getHost() != null
                && configuredServer.getHost().equalsIgnoreCase(uri.getHost());
    }

    private boolean openExternally(Uri uri) {
        try {
            Intent intent = new Intent(Intent.ACTION_VIEW, uri);
            startActivity(intent);
            return true;
        } catch (ActivityNotFoundException exception) {
            Toast.makeText(this, "이 링크를 열 수 있는 앱이 없습니다.", Toast.LENGTH_LONG).show();
            return true;
        }
    }

    private void showError(String title, String message) {
        progressBar.setVisibility(View.GONE);
        errorTitle.setText(title);
        errorMessage.setText(message);
        errorPanel.setVisibility(View.VISIBLE);
    }

    private void hideError() {
        errorPanel.setVisibility(View.GONE);
    }

    private String describeSslError(SslError error) {
        return switch (error.getPrimaryError()) {
            case SslError.SSL_EXPIRED -> "인증서가 만료되었습니다.";
            case SslError.SSL_IDMISMATCH -> "인증서의 호스트 이름이 서버 주소와 다릅니다.";
            case SslError.SSL_NOTYETVALID -> "인증서의 유효 기간이 아직 시작되지 않았습니다.";
            case SslError.SSL_UNTRUSTED -> "신뢰할 수 없는 인증기관이 발급한 인증서입니다.";
            case SslError.SSL_DATE_INVALID -> "인증서 날짜가 올바르지 않습니다.";
            case SslError.SSL_INVALID -> "인증서가 유효하지 않습니다.";
            default -> "인증서를 확인할 수 없습니다.";
        };
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode != FILE_CHOOSER_REQUEST || pendingFileCallback == null) {
            return;
        }
        Uri[] result = WebChromeClient.FileChooserParams.parseResult(resultCode, data);
        pendingFileCallback.onReceiveValue(result);
        pendingFileCallback = null;
    }

    @Override
    public void onBackPressed() {
        if (webView.canGoBack()) {
            webView.goBack();
        } else {
            super.onBackPressed();
        }
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        webView.saveState(outState);
        super.onSaveInstanceState(outState);
    }

    @Override
    protected void onDestroy() {
        if (pendingFileCallback != null) {
            pendingFileCallback.onReceiveValue(null);
            pendingFileCallback = null;
        }
        webView.stopLoading();
        webView.setWebChromeClient(null);
        webView.setWebViewClient(null);
        webView.destroy();
        super.onDestroy();
    }

    private final class FileExploreChromeClient extends WebChromeClient {
        @Override
        public void onProgressChanged(WebView view, int newProgress) {
            progressBar.setProgress(newProgress);
            progressBar.setVisibility(newProgress >= 100 ? View.GONE : View.VISIBLE);
        }

        @Override
        public boolean onShowFileChooser(
                WebView webView,
                ValueCallback<Uri[]> filePathCallback,
                FileChooserParams fileChooserParams
        ) {
            if (pendingFileCallback != null) {
                pendingFileCallback.onReceiveValue(null);
            }
            pendingFileCallback = filePathCallback;
            try {
                Intent chooserIntent = fileChooserParams.createIntent();
                startActivityForResult(chooserIntent, FILE_CHOOSER_REQUEST);
                return true;
            } catch (ActivityNotFoundException exception) {
                pendingFileCallback = null;
                Toast.makeText(MainActivity.this, "파일 선택기를 열 수 없습니다.", Toast.LENGTH_LONG).show();
                return false;
            }
        }
    }

    private final class FileExploreWebViewClient extends WebViewClient {
        @Override
        public void onPageStarted(WebView view, String url, Bitmap favicon) {
            hideError();
            progressBar.setVisibility(View.VISIBLE);
        }

        @Override
        public void onPageFinished(WebView view, String url) {
            progressBar.setVisibility(View.GONE);
            CookieManager.getInstance().flush();
        }

        @Override
        public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
            Uri uri = request.getUrl();
            String scheme = uri.getScheme() == null ? "" : uri.getScheme().toLowerCase(Locale.ROOT);
            if (scheme.equals("http") || scheme.equals("https")) {
                return belongsToConfiguredServer(uri) ? false : openExternally(uri);
            }
            if (scheme.equals("blob") || scheme.equals("data") || scheme.equals("about")) {
                return false;
            }
            if (scheme.equals("intent")) {
                try {
                    Intent intent = Intent.parseUri(uri.toString(), Intent.URI_INTENT_SCHEME);
                    startActivity(intent);
                    return true;
                } catch (URISyntaxException | ActivityNotFoundException exception) {
                    return true;
                }
            }
            return openExternally(uri);
        }

        @Override
        public void onReceivedError(WebView view, WebResourceRequest request, WebResourceError error) {
            if (!request.isForMainFrame()) {
                return;
            }
            String detail = error.getDescription() == null
                    ? "서버와 통신할 수 없습니다."
                    : error.getDescription().toString();
            showError(
                    "FileServer에 연결할 수 없습니다",
                    detail + "\n\n같은 Wi-Fi인지, 서버 주소와 포트가 맞는지 확인하세요."
            );
        }

        @Override
        public void onReceivedSslError(WebView view, SslErrorHandler handler, SslError error) {
            Uri failingUri = Uri.parse(error.getUrl());
            String host = failingUri.getHost();
            if (host != null && host.equalsIgnoreCase(temporarilyTrustedSslHost)) {
                handler.proceed();
                return;
            }

            handler.cancel();
            String reason = describeSslError(error);
            new AlertDialog.Builder(MainActivity.this)
                    .setTitle("안전하지 않은 인증서")
                    .setMessage(
                            reason
                                    + "\n\n서버: " + error.getUrl()
                                    + "\n\n내부망의 직접 만든 FileServer가 맞는지 확인한 경우에만 계속하세요. "
                                    + "이 허용은 앱을 종료할 때까지만 유지됩니다."
                    )
                    .setNegativeButton("취소", (dialog, which) -> showError(
                            "인증서 오류",
                            reason + "\n\n신뢰되는 인증서를 설치하거나 서버 설정을 확인하세요."
                    ))
                    .setPositiveButton("이번 실행에서 계속", (dialog, which) -> {
                        temporarilyTrustedSslHost = host;
                        webView.reload();
                    })
                    .show();
        }
    }
}

File: FileExploreAndroid/app/src/main/java/com/syj5385/fileexplore/MainActivity.kt
package com.syj5385.fileexplore

import android.Manifest
import android.app.Activity
import android.app.AlertDialog
import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.net.Uri
import android.net.http.SslCertificate
import android.net.http.SslError
import android.os.Build
import android.os.Bundle
import android.text.InputType
import android.view.View
import android.view.ViewGroup
import android.view.WindowInsets
import android.webkit.CookieManager
import android.webkit.DownloadListener
import android.webkit.HttpAuthHandler
import android.webkit.SslErrorHandler
import android.webkit.ValueCallback
import android.webkit.WebChromeClient
import android.webkit.WebResourceError
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.Button
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast

class MainActivity : Activity() {

    companion object {
        private const val PREFS_NAME = "file_explore_preferences"
        private const val PREF_SERVER_URL = "server_url"
        private const val FILE_CHOOSER_REQUEST = 1001
        private const val DOWNLOAD_PERMISSION_REQUEST = 1002
    }

    private lateinit var webView: WebView
    private lateinit var pageProgress: ProgressBar
    private lateinit var pageTitle: TextView
    private var filePathCallback: ValueCallback<Array<Uri>>? = null
    private var pendingDownload: FileDownloadRequest? = null
    private var homeUrl: String? = null
    private var acceptedSslHost: String? = null
    private var acceptedSslCertificateDer: ByteArray? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        webView = findViewById(R.id.webView)
        pageProgress = findViewById(R.id.pageProgress)
        pageTitle = findViewById(R.id.pageTitle)

        applySystemBarInsets()
        configureWebView()

        findViewById<Button>(R.id.reloadButton).setOnClickListener {
            if (webView.url.isNullOrBlank()) {
                homeUrl?.let(webView::loadUrl)
            } else {
                webView.reload()
            }
        }

        findViewById<Button>(R.id.serverButton).setOnClickListener {
            showServerDialog(required = false)
        }

        if (savedInstanceState != null) {
            webView.restoreState(savedInstanceState)
            homeUrl = preferences().getString(PREF_SERVER_URL, null)
            updateTitle(webView.url ?: homeUrl)
        } else {
            val savedUrl = preferences().getString(PREF_SERVER_URL, null)
            if (savedUrl.isNullOrBlank()) {
                showServerDialog(required = true)
            } else {
                loadServer(savedUrl)
            }
        }
    }

    private fun configureWebView() {
        CookieManager.getInstance().apply {
            setAcceptCookie(true)
            setAcceptThirdPartyCookies(webView, true)
        }

        val isDebuggable =
            (applicationInfo.flags and ApplicationInfo.FLAG_DEBUGGABLE) != 0
        WebView.setWebContentsDebuggingEnabled(isDebuggable)

        webView.settings.apply {
            javaScriptEnabled = true
            domStorageEnabled = true
            databaseEnabled = true
            loadsImagesAutomatically = true
            useWideViewPort = true
            loadWithOverviewMode = true
            builtInZoomControls = false
            displayZoomControls = false
            mediaPlaybackRequiresUserGesture = true
            allowFileAccess = false
            allowContentAccess = true
            setSupportMultipleWindows(false)
            userAgentString = "$userAgentString FileExploreAndroid/1.0.3"
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                safeBrowsingEnabled = true
            }
        }

        webView.webViewClient = object : WebViewClient() {
            override fun onPageStarted(view: WebView?, url: String?, favicon: Bitmap?) {
                pageProgress.visibility = View.VISIBLE
                updateTitle(url)
            }

            override fun onPageFinished(view: WebView?, url: String?) {
                pageProgress.visibility = View.GONE
                CookieManager.getInstance().flush()
                updateTitle(view?.title ?: url)
            }

            override fun shouldOverrideUrlLoading(
                view: WebView?,
                request: WebResourceRequest?
            ): Boolean = handleNavigation(request?.url)

            @Suppress("DEPRECATION")
            override fun shouldOverrideUrlLoading(view: WebView?, url: String?): Boolean {
                return handleNavigation(url?.let(Uri::parse))
            }

            override fun onReceivedError(
                view: WebView?,
                request: WebResourceRequest?,
                error: WebResourceError?
            ) {
                if (request?.isForMainFrame == true) {
                    pageProgress.visibility = View.GONE
                    val message = error?.description?.toString().orEmpty()
                    if (message.isNotBlank()) {
                        Toast.makeText(this@MainActivity, message, Toast.LENGTH_LONG).show()
                    }
                }
            }

            override fun onReceivedSslError(
                view: WebView?,
                handler: SslErrorHandler?,
                error: SslError?
            ) {
                if (handler == null) return
                if (isFinishing || isDestroyed) {
                    handler.cancel()
                    return
                }

                val host = error?.url?.let { Uri.parse(it).host }.orEmpty()
                val certificateDer = sslCertificateDer(error?.certificate)
                val alreadyAccepted = host.equals(acceptedSslHost, ignoreCase = true) &&
                    certificateDer != null &&
                    acceptedSslCertificateDer?.contentEquals(certificateDer) == true
                if (alreadyAccepted) {
                    handler.proceed()
                    return
                }

                val reason = sslErrorDescription(error)
                AlertDialog.Builder(this@MainActivity)
                    .setTitle(R.string.ssl_warning_title)
                    .setMessage(
                        "$host\n$reason\n\n" +
                            "계속하면 이번 앱 실행 동안 이 서버와 동일한 인증서에만 접속과 다운로드를 허용합니다."
                    )
                    .setNegativeButton(R.string.ssl_warning_cancel) { _, _ -> handler.cancel() }
                    .setPositiveButton(R.string.ssl_warning_continue) { _, _ ->
                        acceptedSslHost = host
                        acceptedSslCertificateDer = certificateDer?.copyOf()
                        handler.proceed()
                    }
                    .setOnCancelListener { handler.cancel() }
                    .show()
            }

            override fun onReceivedHttpAuthRequest(
                view: WebView?,
                handler: HttpAuthHandler?,
                host: String?,
                realm: String?
            ) {
                if (handler == null) return
                showHttpAuthDialog(handler, host.orEmpty(), realm.orEmpty())
            }
        }

        webView.webChromeClient = object : WebChromeClient() {
            override fun onProgressChanged(view: WebView?, newProgress: Int) {
                pageProgress.progress = newProgress
                pageProgress.visibility = if (newProgress in 0..99) View.VISIBLE else View.GONE
            }

            override fun onReceivedTitle(view: WebView?, title: String?) {
                updateTitle(title ?: view?.url)
            }

            override fun onShowFileChooser(
                webView: WebView?,
                filePathCallback: ValueCallback<Array<Uri>>?,
                fileChooserParams: FileChooserParams?
            ): Boolean {
                this@MainActivity.filePathCallback?.onReceiveValue(null)
                this@MainActivity.filePathCallback = filePathCallback

                val chooserIntent = try {
                    fileChooserParams?.createIntent() ?: Intent(Intent.ACTION_OPEN_DOCUMENT).apply {
                        addCategory(Intent.CATEGORY_OPENABLE)
                        type = "*/*"
                        putExtra(Intent.EXTRA_ALLOW_MULTIPLE, true)
                    }
                } catch (_: Exception) {
                    Intent(Intent.ACTION_OPEN_DOCUMENT).apply {
                        addCategory(Intent.CATEGORY_OPENABLE)
                        type = "*/*"
                        putExtra(Intent.EXTRA_ALLOW_MULTIPLE, true)
                    }
                }

                return try {
                    startActivityForResult(chooserIntent, FILE_CHOOSER_REQUEST)
                    true
                } catch (_: ActivityNotFoundException) {
                    this@MainActivity.filePathCallback = null
                    Toast.makeText(
                        this@MainActivity,
                        R.string.file_chooser_failed,
                        Toast.LENGTH_LONG
                    ).show()
                    false
                }
            }
        }

        webView.setDownloadListener(DownloadListener { url, userAgent, contentDisposition, mimeType, _ ->
            startDownload(url, userAgent, contentDisposition, mimeType)
        })
    }

    private fun loadServer(url: String) {
        val normalized = normalizeServerUrl(url) ?: return
        homeUrl = normalized
        preferences().edit().putString(PREF_SERVER_URL, normalized).apply()
        updateTitle(normalized)
        webView.loadUrl(normalized)
    }

    private fun showServerDialog(required: Boolean) {
        val input = EditText(this).apply {
            hint = getString(R.string.server_dialog_hint)
            inputType = InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_URI
            setSingleLine(true)
            setText(homeUrl ?: preferences().getString(PREF_SERVER_URL, ""))
            setSelection(text.length)
        }

        val horizontalPadding = (20 * resources.displayMetrics.density).toInt()
        val container = LinearLayout(this).apply {
            orientation = LinearLayout.VERTICAL
            setPadding(horizontalPadding, 0, horizontalPadding, 0)
            addView(
                input,
                LinearLayout.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.WRAP_CONTENT
                )
            )
        }

        val dialog = AlertDialog.Builder(this)
            .setTitle(R.string.server_dialog_title)
            .setView(container)
            .setPositiveButton(R.string.connect, null)
            .apply {
                if (!required) {
                    setNegativeButton(R.string.cancel, null)
                }
            }
            .setCancelable(!required)
            .create()

        dialog.setOnShowListener {
            dialog.getButton(AlertDialog.BUTTON_POSITIVE).setOnClickListener {
                val normalized = normalizeServerUrl(input.text.toString())
                if (normalized == null) {
                    input.error = getString(R.string.invalid_server_url)
                    return@setOnClickListener
                }
                dialog.dismiss()
                acceptedSslHost = null
                acceptedSslCertificateDer = null
                webView.stopLoading()
                webView.clearHistory()
                loadServer(normalized)
            }
        }

        dialog.show()
        input.requestFocus()
    }

    private fun normalizeServerUrl(raw: String): String? {
        val trimmed = raw.trim()
        if (trimmed.isEmpty()) return null

        val withScheme = if (trimmed.startsWith("http://", ignoreCase = true) ||
            trimmed.startsWith("https://", ignoreCase = true)
        ) {
            trimmed
        } else {
            "https://$trimmed"
        }

        return try {
            val uri = Uri.parse(withScheme)
            if ((uri.scheme == "http" || uri.scheme == "https") && !uri.host.isNullOrBlank()) {
                withScheme.trimEnd('/')
            } else {
                null
            }
        } catch (_: Exception) {
            null
        }
    }

    private fun handleNavigation(uri: Uri?): Boolean {
        if (uri == null) return false
        return when (uri.scheme?.lowercase()) {
            "http", "https" -> false
            else -> {
                try {
                    startActivity(Intent(Intent.ACTION_VIEW, uri))
                } catch (_: ActivityNotFoundException) {
                    Toast.makeText(this, R.string.open_external_failed, Toast.LENGTH_LONG).show()
                }
                true
            }
        }
    }

    private fun startDownload(
        url: String?,
        userAgent: String?,
        contentDisposition: String?,
        mimeType: String?
    ) {
        if (url.isNullOrBlank() || url.startsWith("blob:", ignoreCase = true)) {
            Toast.makeText(this, R.string.download_failed, Toast.LENGTH_LONG).show()
            return
        }

        val download = FileDownloadRequest(
            url = url,
            userAgent = userAgent,
            contentDisposition = contentDisposition,
            mimeType = mimeType,
            cookie = CookieManager.getInstance().getCookie(url),
            trustedHost = acceptedSslHost,
            trustedCertificateDer = acceptedSslCertificateDer?.copyOf()
        )
        if (requiresLegacyStoragePermission() &&
            checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE) !=
            PackageManager.PERMISSION_GRANTED
        ) {
            pendingDownload = download
            requestPermissions(
                arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE),
                DOWNLOAD_PERMISSION_REQUEST
            )
            return
        }

        enqueueDownload(download)
    }

    private fun requiresLegacyStoragePermission(): Boolean {
        return Build.VERSION.SDK_INT <= Build.VERSION_CODES.P
    }

    private fun enqueueDownload(download: FileDownloadRequest) {
        FileDownloader.enqueue(applicationContext, download)
    }

    private fun sslCertificateDer(certificate: SslCertificate?): ByteArray? {
        if (certificate == null) return null
        return runCatching {
            val state = SslCertificate.saveState(certificate) ?: return null
            state.getByteArray("x509-certificate")?.copyOf()
        }.getOrNull()
    }

    private fun showHttpAuthDialog(handler: HttpAuthHandler, host: String, realm: String) {
        val username = EditText(this).apply {
            hint = getString(R.string.username)
            setSingleLine(true)
        }
        val password = EditText(this).apply {
            hint = getString(R.string.password)
            inputType = InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_PASSWORD
            setSingleLine(true)
        }
        val padding = (20 * resources.displayMetrics.density).toInt()
        val fields = LinearLayout(this).apply {
            orientation = LinearLayout.VERTICAL
            setPadding(padding, 0, padding, 0)
            addView(username)
            addView(password)
        }

        val details = listOf(host, realm).filter { it.isNotBlank() }.joinToString(" · ")
        AlertDialog.Builder(this)
            .setTitle(R.string.http_auth_title)
            .setMessage(details)
            .setView(fields)
            .setNegativeButton(R.string.cancel) { _, _ -> handler.cancel() }
            .setPositiveButton(R.string.login) { _, _ ->
                handler.proceed(username.text.toString(), password.text.toString())
            }
            .setOnCancelListener { handler.cancel() }
            .show()
    }

    private fun sslErrorDescription(error: SslError?): String {
        return when (error?.primaryError) {
            SslError.SSL_DATE_INVALID -> "인증서 날짜가 올바르지 않습니다."
            SslError.SSL_EXPIRED -> "인증서가 만료되었습니다."
            SslError.SSL_IDMISMATCH -> "서버 주소와 인증서 이름이 일치하지 않습니다."
            SslError.SSL_NOTYETVALID -> "인증서가 아직 유효하지 않습니다."
            SslError.SSL_UNTRUSTED -> "자체 서명되었거나 신뢰되지 않은 인증서입니다."
            else -> "TLS 인증서 검증에 실패했습니다."
        }
    }

    private fun updateTitle(value: String?) {
        val text = value.orEmpty().trim()
        if (text.isBlank()) {
            pageTitle.setText(R.string.app_name)
            return
        }

        val host = runCatching { Uri.parse(text).host }.getOrNull()
        pageTitle.text = host?.takeIf { it.isNotBlank() } ?: text
    }

    private fun preferences() = getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    private fun applySystemBarInsets() {
        val root = findViewById<View>(R.id.root)
        root.setOnApplyWindowInsetsListener { view, insets ->
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                val bars = insets.getInsets(WindowInsets.Type.systemBars())
                view.setPadding(bars.left, bars.top, bars.right, bars.bottom)
            } else {
                @Suppress("DEPRECATION")
                view.setPadding(
                    insets.systemWindowInsetLeft,
                    insets.systemWindowInsetTop,
                    insets.systemWindowInsetRight,
                    insets.systemWindowInsetBottom
                )
            }
            insets
        }
        root.requestApplyInsets()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == FILE_CHOOSER_REQUEST) {
            val result = WebChromeClient.FileChooserParams.parseResult(resultCode, data)
            filePathCallback?.onReceiveValue(result)
            filePathCallback = null
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode != DOWNLOAD_PERMISSION_REQUEST) return

        val download = pendingDownload
        pendingDownload = null
        if (grantResults.firstOrNull() == PackageManager.PERMISSION_GRANTED && download != null) {
            enqueueDownload(download)
        } else {
            Toast.makeText(this, R.string.download_permission_denied, Toast.LENGTH_LONG).show()
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        webView.saveState(outState)
        super.onSaveInstanceState(outState)
    }

    override fun onPause() {
        CookieManager.getInstance().flush()
        webView.onPause()
        super.onPause()
    }

    override fun onResume() {
        super.onResume()
        webView.onResume()
    }

    override fun onDestroy() {
        pendingDownload = null
        filePathCallback?.onReceiveValue(null)
        filePathCallback = null
        webView.stopLoading()
        webView.webChromeClient = null
        webView.webViewClient = WebViewClient()
        webView.destroy()
        super.onDestroy()
    }

    @Suppress("DEPRECATION")
    override fun onBackPressed() {
        if (webView.canGoBack()) {
            webView.goBack()
        } else {
            super.onBackPressed()
        }
    }
}
